tests/core/fs/workspacefs/test_workspace_fs.py

```diff
@@ -438,3 +441,77 @@ async def mockup(*args):
     with running_backend.offline():
         with pytest.raises(FSBackendOfflineError):
             await alice_workspace.get_reencryption_need()
+
+
+@pytest.mark.trio
+async def test_upload_blocks(alice_user_fs, alice2_user_fs, running_backend, monkeypatch):
+    wid = await create_shared_workspace("w", alice_user_fs, alice2_user_fs)
+    alice_workspace = alice_user_fs.get_workspace(wid)
+
+    alice2_workspace = alice2_user_fs.get_workspace(wid)
+
+    fspath = "/taz"
+
+    await alice_workspace.touch(fspath)
+    await alice_workspace.sync()
+
+    await alice_workspace.write_bytes(fspath, b"aaaaaa" * DEFAULT_BLOCK_SIZE)
+
+    info = await alice_workspace.path_info(fspath)
+
+    assert info["base_version"] == 1
+    assert info["is_placeholder"] is False
+    assert info["need_sync"] is True
+
+    assert await alice2_workspace.read_bytes(fspath) == bytearray()
+
+    # mock the synchronization_step method to retrieve the new manifest
+    vanilla_transaction_synchronization_step = alice_workspace.transactions.synchronization_step
+    block_id_error = None
+
+    async def mock_synchronization_step(entry_id, *args, **kwargs):
+        nonlocal block_id_error
+        nonlocal info
+        res = await vanilla_transaction_synchronization_step(entry_id, *args, **kwargs)
+        if entry_id == info["id"] and res:
+            block_id_error = res.blocks[2].id
+
+        return res
+
+    monkeypatch.setattr(
+        alice_workspace.transactions, "synchronization_step", mock_synchronization_step
+    )
+
+    vanilla_backend_block_create = running_backend.backend.block.create
+
+    async def mock_backend_block_create(*args, **kwargs):
+        nonlocal block_id_error
+        if kwargs["block_id"] == block_id_error:
+            raise BlockNotFoundError()
+        return await vanilla_backend_block_create(*args, **kwargs)
+
+    monkeypatch.setattr(running_backend.backend.block, "create", mock_backend_block_create)
+
+    with pytest.raises(FSError):
+        await alice_workspace.sync()
+
+    info = await alice_workspace.path_info(fspath)
+
+    assert info["base_version"] == 1
+    assert info["is_placeholder"] is False
+    assert info["need_sync"] is True
+
+    monkeypatch.setattr(running_backend.backend.block, "create", vanilla_backend_block_create)
+
+    await alice2_workspace.sync()
```

Comment:
```python
# Given sync has failed, alice2 should still see file in v1
```
Besides I think you should keep this test before the unmock (checking alice2 view is part of the checks done on the view of alice)
And you should use path_info instead of read_bytes (given it gives you the size attribute that sould be enough:

```python
with pytest.raises(FSError):
    await alice_workspace.sync()

# File sync has failed, so it info should be the same
info = await alice_workspace.path_info(fspath)
assert info["base_version"] == 1
assert info["need_sync"] is True
# Other device doesn't see the modifications that have failed to be synced
await alice2_workspace.sync()
info = await alice2_workspace.path_info(fspath)
assert info["base_version"] == 1
assert info["need_sync"] is False
assert info["size"] == 0
```
(note I remove the check on `is_placeholder` to be consistent with the "avoid this corner-case in the rest of the test" comment I made above 😄 )